### ISSUE TITLE
[wgsl] Proposal: Remove pointer out parameters from modf, frexp

### ISSUE THREAD

**Issue by ben-clayton** [gpuweb/gpuweb#1480]
These two intrinsics are the only two in the spec that have a pointer output parameter:

```rust
modf(e1: T, e2: ptr<T> ) -> T
modf(e1: vecN<T>, e2: ptr<vecN<T>>) -> vecN<T>
frexp(e1: T, e2: ptr<I> ) -> T
frexp(e1: vecN<T>, e2: ptr<vecN<I>>) -> vecN<T>
```

The pointer parameter makes usage a bit clunky, and we're still debating pointers in the language.
We might want to consider changing their signature to return a builtin anonymous structure:

```rust
modf(e: f32) -> struct {
  fractional : f32;
  whole : f32;
};

modf(e: vecN<f32>) -> struct {
  fractional : vecN<f32>;
  whole : vecN<f32>;
};

frexp(e: f32) -> struct {
  significand : f32;
  exponent : i32;
};

frexp(e: vecN<f32>) -> struct {
  significand : vecN<f32>;
  exponent : vecN<i32>;
};
```

As the returned structure is anonymous (and likely needs to be due to the template types), we'd need type inference to be able to declare a variable for storage.

Alternatively, we could unroll the templates:

```rust
// Predefined by the WGSL language.
struct modf_result_f32 {
  fractional : f32;
  whole : f32;
};
struct modf_result_vec2_f32 {
  fractional : vec2<f32>;
  whole : vec2<f32>;
};
struct modf_result_vec3_f32 {
  fractional : vec3<f32>;
  whole : vec3<f32>;
};
struct modf_result_vec4_f32 {
  fractional : vec4<f32>;
  whole : vec4<f32>;
};
struct frexp_result_f32 {
  significand : f32;
  exponent : i32;
};
struct frexp_result_vec2_f32 {
  significand : vec2<f32>;
  exponent : vec2<i32>;
};
struct frexp_result_vec3_f32 {
  significand : vec3<f32>;
  exponent : vec3<i32>;
};
struct frexp_result_vec4_f32 {
  significand : vec4<f32>;
  exponent : vec4<i32>;
};

modf(e: f32) -> modf_result_f32
modf(e: vec2<f32>) -> modf_result_vec2_f32
modf(e: vec3<f32>) -> modf_result_vec3_f32
modf(e: vec4<f32>) -> modf_result_vec4_f32
frexp(e: f32) -> frexp_result_f32
frexp(e: vec2<f32>) -> frexp_result_vec2_f32
frexp(e: vec3<f32>) -> frexp_result_vec3_f32
frexp(e: vec4<f32>) -> frexp_result_vec4_f32
```

This would allow us to implement this today, but does litter the global namespace.

**Comment by ben-clayton** [comment]
Now that we have type inferencing (for `let`, I'm assuming `var` soon), how does returning a builtin **anonymous** structure now sound?
It means you'll have to consume or copy the values out after the call, but I don't think that's much of an issue. If we find we want to add a named structure for these (templated or not) at a later date, we can easily do so without breaking any existing code.

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-04-13-Minutes#wgsl-proposal-remove-pointer-out-parameters-from-modf-frexp--issue-1480--gpuwebgpuweb">WGSL meeting minutes 2021-04-13</a></summary>




*   DM: Do we want to spend innovation budget on anonymous structs here?
*   BC: Previously there were concerns about having to have a named struct to return here, but now with type inference, we could return anonymous structs here. Not necessarily proposing userland anonymous structs, and if we decided to name them in the future, we could name them.
*   DM: Some complexity, new types for these builtins.
*   AB: They need special types named or not, if we’re returning structs.
*   BC: Motivation: In the future, we will probably want this with generics.
*   DM: It might be nice to do this with tuples instead, and we can’t switch to tuples later here.
*   BC: I think named fields are nicer than indexing into tuples
*   KN: +1
*   JG: We can still switch to tuples if we spec them before we commit to backwards compat.
*   DM: I feel like named fields is notably different than what we have for normal call-sites, which don’t have named argos
*   AB: Why tuples?
*   DM: Nicer for wrapping? Vs nested structures
*   KN: Sort of only difference between anon structs and tuples is that anon structs are more strictly typed, and cannot be assigned to a different anon struct, unless you have “structural typing”, where what matters is the shape of the struct and not its name.
*   DM: Did we consider implementation cost?
*   BC: Probably less work than having to declare many structures here
*   DM: I see tuples are being first-class constructs
*   BC: I think anon structs can be pretty similar to what you see for tuples.
*   JG: Can we use this to first solve the multiple-return-value problem? Vs what we have now with part-return-part-output-pointer. 
*   BC: Yeah, that’s why I proposed it
*   AB: That’s why I’m bearish on this: We have a solution that works, and this would require something new.
*   BC: I don’t care as much as I used to, but I feel like if we don’t make it good now, it will always suck.
*   AB: We could add overloads later, but I don’t want to get too far off in additions to the language today, just for this.
*   KN: I don’t think named structs are a dealbreaker
*   DM: Yeah, maybe there’s no real implementation issue here, but I wondered if you’d tried it yet. We’ll need to prototype.
*   JG: Is named structs that bad here? No new feature needed, just boilerplate.
*   DM: Would we want to follow up with a general anon struct proposal anyway?
*   JG: Like tuples, proposals welcome
*   AB: We didn’t agree on names yet, which we need for speccing
*   JG: Probably just try to spec something, as long as we don’t think naming discussions are non-consensable.
*   BC: This proposal was because we would not need to agree on names! :) What we have now is workable, but no strong feelings.
*   DM: I like anon structs better than named structs.
*   BC: I was thinking anon structs would be easy to plug into the lang. KN said though that this adds a foible that structure types are based on names, which becomes maybe problematic for anon structs. 
*   JG
*   BC: What if you have two calls to `modf`, but do they have the same type?
*   JB: Probably want that to work, and there is prior art here.
*   BC: If we wanted anon structs, we need this struct duck-typing here
*   DM: If we wanted to name them later, we wouldn’t be able to s/anon/named/, because programs would break.
*   AB: Does structural equivalence consider name or just type?
*   JG: Can we just name them and talk about anon struct proposals right here right now?
*   AB: Some reservations allowing anon structs for users



</details>

**Comment by dneto0** [comment]
> Now that we have type inferencing (for let, I'm assuming var soon), how does returning a builtin anonymous structure now sound?

Sounds good.   But I think you still need member names?  We either make custom ones per function, or 
need a convention, like  `f0`, `f1` etc.

          let result = frexp(12.0);
          let fraction = result.f0;
          let exponent = result.f1;

Aside:  OpenCL C went with  a swizzle syntax where 's' was the prefix then you appended a bunch of hex digits. So if v was a float16, then  v.s0a was a 2 element float vector with the first and 11th elements of v.
See https://www.khronos.org/registry/OpenCL/specs/3.0-unified/html/OpenCL_C.html#vector-components 

We already have types in the type system that can't be spelled out in the program. (e.g. reference types).  So having an anonymous struct for specific purposes seems ok and within the broader framework.

**Comment by ben-clayton** [comment]
Yes, sorry, I didn't mean to give the impression the struct wouldn't have member names. IMO, that's the compelling argument over tuples. See [the OP](https://github.com/gpuweb/gpuweb/issues/1480#issue-819957286) for my suggested names, which are intentionally descriptive.

I remember we discussed the topic of anonymous structs at some length in the WG meeting - the one thing we discussed but didn't resolve was the rules about how anonymous structures would behave W.R.T type equality, and so value assignability.

i.e. is this legal, or not?

```
var a = frexp(1.0);
var b = frexp(2.0);
a = b; // error or legal?
```

Right now, the overloads of `frexp` and `modf` have no common returned structures (identical member field types), so that simplifies things a bit. However, if added another intrinsic that did return an anonymous structure with the same sequence of field types, but with different member names, should they be type-equal or not? (Possibly doesn't have to be answered until we hit this, but would be good to have a general idea on how we'd tackle this)

**Comment by kvark** [comment]
It seems that structures can be very useful even if this assignment is not supported, and we can enable it some time later down the road.

**Comment by ben-clayton** [comment]
I agree, but we still need to decide what the rules are.

**Comment by litherum** [comment]
I listed a bunch of arguments on both sides in https://github.com/gpuweb/gpuweb/issues/1870. And a bit of a digression on possible forward compatibility with tuples.

**Comment by litherum** [comment]
As for picking names, I think the ones that @ben-clayton came up with in the OP are fine.

**Comment by ben-clayton** [comment]
I've been mulling over the discussion around these builtins returning the anonymous structures vs the named builtin structures, and I'm rather strongly opposed to named builtin structures.

Long term, I believe we're in agreement that an anonymous structure or tuple is preferable over a fully permutated list of named return structure types. I'm concerned that we have no clear way to migrate away from these named structures once they land in the spec.

My primary concerns are:

1. Once we have `modf` and `frexp` return a named structure, how can we change the spec to return anything different? The only approach I could imagine is adding some hacky special rules to the spec that states that these special return types can be implicitly cast to the new-thing.
2. If we have `modf` and `frexp` return these named types for Chrome's WebGPU Origin Trial next month, then there's a good chance we'll have to support these indefinitely, in order to not break existing shader code. This is not something I'd want to maintain in the spec, nor the compiler.
3. Nobody is going to want to write these names. If people are just going to use type inferencing for this, then anonymous structures are no less convenient.

So instead, it would be my preference that:

1. These methods would have the signature:

```rust
modf(e: T) -> struct {
  fractional : f32;
  whole : f32;
};
modf(e: vecN<T>) -> struct {
  fractional : vecN<f32>;
  whole : vecN<f32>;
};
frexp(e: T) -> struct {
  significand : f32;
  exponent : i32;
};
frexp(e: vecN<T>) -> struct {
  significand : vecN<f32>;
  exponent : vecN<i32>;
};
```
2. We **don't** update the spec to allow for user-declared anonymous structure types (yet)
3. We **do** update the spec to say that anonymous structures which have the same sequence of member types are equivalent. This resolves [the assignability question above](https://github.com/gpuweb/gpuweb/issues/1480#issuecomment-863587944), and will allow for users to declare locals / members of these types in the future.

If we went with anonymous structures, then changed our minds, the anonymous structure can easily be turned into a named structure at any later date. The inverse is not true.

**Comment by litherum** [comment]
Python [returns a tuple for `frexp()`](https://docs.python.org/3/library/math.html#math.frexp)

Haskell [does too](https://hackage.haskell.org/package/cmath-0.3/docs/Foreign-C-Math-Double.html)

[So does Go](https://golang.org/src/math/frexp.go)

[And Kotlin too](https://kotlin.github.io/kotlin-numpy/kotlin-numpy/org.jetbrains.numkt.math/frexp.html)

Swift [uses an out param](https://developer.apple.com/documentation/kernel/1557221-frexp)

It looks like Rust doesn't have a `frexp()` function, but the [one that was proposed returns a tuple](https://github.com/rust-lang/rust/issues/27752)

(I'm intentionally not including languages in this list which don't have tuples)

**Comment by ben-clayton** [comment]
In terms of usability, I don't see how:

```rust
var significand, exponent = frexp(f);
blah(significand, exponent);
```

is significantly better than:

```rust
var r = frexp(f);
blah(r.significand, r.exponent);
```

Go doesn't have tuples, it supports multiple return values. The difference being that there is no composite type for the N-return types. Go also supports returning a single value, anonymous structure (although usage is clunky). 

Side note: in Go there's actually no difference between an anonymous structure and a named one - you simply use `type` to give the structure a name. I like this, but WGSL has opted for the structure name to be the unique key for the type declaration, where as in Go, it's the member name & types that make a structure unique.

**Comment by litherum** [comment]
> In terms of usability, I don't see how:
> 
> ```rust
> var significand, exponent = frexp(f);
> blah(significand, exponent);
> ```
> 
> is significantly better than:
> 
> ```rust
> var r = frexp(f);
> blah(r.significand, r.exponent);
> ```
> 

It's better in situations where you want to use `significand` and `exponent` multiple times. You don't have to keep writing all those `r.`s everywhere.

```
var r = frexp(f);
for (var i = 0; i < something; i = i + i) {
    blah(r.significand, r.exponent);
    value = r.significand + r.exponent * something;
    if (f32(i) < r.exponent) {
        doSomething(r.exponent - somethingElse);
    }
}
```

It also allows you to more easily rename the output and use it:

```
var product, exponent = frexp(f);
// No extra variable needed here to rename "significand" to "product"
for (var i = 0; i < something; i = i + i) {
    product = product * something;
}
```

**Comment by litherum** [comment]
I think it's important for `modf` that the fields be named - it's too easy to forget that the first one is the fractional part and the second one is the whole part. But, for `frexp()`, the types of `significand` and `exponent` make it obvious which one is which. So, how about a proposal:

1. Accept anonymous structs, as you have described them here
2. Either every field has a name, or no field has a name
3. If no field has a name, there's a shorthand of `(f32, i32)` (possibly also forbidding the `struct` syntax in this case)
4. The identity of these anonymous structs includes their names (and potentially also whether or not the word "struct" appeared in their declaration) (More on this below.)
5. We allow automatic unwrapping in variable declarations

For 4, this would mean that the following would be illegal:
```
var x: struct {
    significand: f32;
    exponent: i32;
} = ...;
var y = struct {
    textureCoordinate: f32;
    texelOffset: i32;
} = ...;
x = y;
```

I think the above is important to disallow - it's too error prone.

But, the following would be legal:

```
var x: (f32, i32) = ...;
var y: (f32, i32) = ...;
x = y;
```

**Comment by ben-clayton** [comment]
Fair examples, but I personally don't see that `r.` as too much of a burden.

One counter argument I have is that this multiple return values requires the user remembers return value order:

```rust
var frac, whole = modf(x);
```

Would be an easy bug to introduce that is not possible with the anonymous structures.

So am I right in thinking that you consider `tuples` and multiple-return-values as one in the same thing?

**Comment by litherum** [comment]
> So am I right in thinking that you consider tuples and multiple-return-values as one in the same thing?

If you wanted to build a tuple, it seems unfortunate to have to make a whole function to do it (especially because functions can't be nested, which means there's an action-at-a-distance kind of thing going on)

**Comment by ben-clayton** [comment]
We posted at the same time.

I quite like some of the suggestions you've made, but I think this is bringing far more into the spec than we want / need for MVP.

> The identity of these anonymous structs includes their names

I could probably live with this, but I'm not sure I agree with your example of being too error prone.
My main concern with having the names be part of the type decl is that I could imagine some (most?) developers might consider `significand` tediously long and difficult to type, however it's a good descriptive name. By omitting the member names from the type decl, they could easily declare their own short-names for these.

Brackets for short-names and automatic unwrapping are nice-to-haves, and I don't personally dislike them, but are likely contentious and are probably going to take a while before we reach agreement.

How opposed are you to accepting 1. (possibly with names as part of type decl) with continued discussions for the other suggestions you've made?

**Comment by ben-clayton** [comment]
With further consideration, I'm totally fine with having the member names part of the anonymous type declaration.

If we decide that this is too restrictive, we can always drop back to my original proposal of only considering the member types, at any later date, without breaking any existing WGSL. The inverse is not true.

So SGTM

**Comment by kvark** [comment]
I think your proposal of having these functions return unnamed structures is very good. It doesn't need to introduce any new syntax or rules. It's better than tuples because the names allow to carry important semantics.

We don't have to invest into the whole "let's expose structural typing to users!" right now. That would be a huge distraction from shipping.

At a later point, we'll have a choice:
  - either actually name those structures and drop any innovation here
  - or expose this ability to users, potentially enhancing the ways to define entry points (which may currently be a little inconvenient once you need to go from 1 output to multiple).
  - or even just keep the structural types only used in some of the standard library methods. As I mentioned on the call, every language has some things that are special and not available to users. In Go it's generics. In Rust it's the `Box` type and other things. It doesn't have to be idealistic.
  - 
The problem of "but I can't name a type!" is not really something users will complain about. They can always copy stuff out in their own types, if needed.

So I strongly feel towards just accepting this proposal and moving forward with it. Thanks @ben-clayton for pushing this!

**Comment by dneto0** [comment]
I endorse everything @kvark said just above.

I'm very open to discussing @litherum's proposed directions, but as a subsequent step.  

FYI. Perl has flexible destructuring that's in line with what the other languages have, without using an explicit tuple type.  Instead, Perl's basic list ("array") does the job:

```perl
use strict;
sub foo { return (1,2,3); }
my ($a,$b,$c) = foo();    # Like $a = 1; $b = 2; $c = 3;
my @triple = foo();       # Like @pair = (1,2,3).
my ($car, @cdr) = foo();  # Like $car = 1; @cdr = (2,3);

```

**Comment by jimblandy** [comment]
WGSL meeting notes:

MM: Can authors use struct returns as well?
AB: Initially limited
RM: What’s the advantage of this approach to begin with?
BC: A named structure would entail declaring a bunch of structures in global type scope. But this locks us into that; we can’t change the builtins to return anything different, even when they’re overloaded. The benefit of anonymous structures is that they give you names for the things, it avoids pointers, and anonymous structures may work for the cases we’ve discussed for multiple return values. But the goal was to not lock us into a different horribly-named global struct for each overload variant.
JG: Does it lock us in? Can’t we just forbid them from naming the type?
BC: That’s an anonymous struct.
JG: But it’s not so hard to specify. If we think it’s easy enough to add anonymous structs, and it’s not too hard to implement, we should just do that
MM: So you’d like to have structs named ModF_f32_f32, but authors can’t name those structs?
JG: yes
AB: Maybe for MVP, we can leave this as pointers. We spent a lot of time on this, and all the other languages take out parameters
MM: We could split these functions into two functions, each of which returns one component
RM: But wouldn’t that be inefficient?
MM: In HLSL, some of these have to be polyfilled anyway. So it wouldn’t necessarily be a performance loss to make them separate functions
AB: [mentions a case where there would be performance loss]
DM: Calling the names out in the spec now is useless. Even if they’re reserved words, new overloads would add new reserved words, breaking older code
JG: We would reserve a prefix, so additions won’t break existing code.
BC: We already have a reserved prefix
DM:
AB: There’s a lot of prior art for reserved prefixes.
JG: even outvars aren’t the end of the world, we can add struct-returning overloads later
JG: Couldn’t struct returns be post-MVP?
MM: Having named structs, reserved names or not, is the expedient thing to do at this point.
RM: Keeping it as-is would be fine for me. It doesn’t require adding anything, it’s ugly but it’s similar to other languages, so it’s not a barrier to adoption
MM: I surveyed a lot of non-shader languages, and all but one of them use outparams [JB: Did I get that right?]
MM: These functions are rare, it’s probably not worth having all this discussion.
AB: Do we want to restrict what storage classes we can use with these outparam pointers?
DM: In other cases, we’ve just exposed what the back end languages do, but we agreed that we should try to clean things up when the back end languages made mistakes. Now we even have the reserved prefix approach, so why aren’t we making the improvement?
AB: Because we don’t agree, we have a workable solution for the MVP, and we have a way to fix it post-MVP.
DM: Does anyone object to returning named structs with a reserved prefix? It seems like everyone liked it
JG: fatigue?
RM: I’m happy with that (reserved prefix)
JG: Okay, let’s do it
BC: I’ll draft a PR that uses a leading underscore for the struct name
MM: When people use the word ‘reserved’, that can mean two things:

- One says you can’t declare a new type in that class

- Another says you can’t even mention things in that class

BC: We want the second case, so that we can change the names later. I’ll make sure we’re getting the second case when I draft the PR
consensus \o/

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-07-20-Minutes#wgsl-proposal-remove-pointer-out-parameters-from-modf-frexp--issue-1480">WGSL meeting minutes 2021-07-20</a></summary>



* MM: Can authors use struct returns as well?
* AB: Initially limited
* RM: What’s the advantage of this approach to begin with?
* BC: A named structure would entail declaring a bunch of structures in global type scope. But this locks us into that; we can’t change the builtins to return anything different, even when they’re overloaded. The benefit of anonymous structures is that they give you names for the things, it avoids pointers, and anonymous structures may work for the cases we’ve discussed for multiple return values. But the goal was to not lock us into a different horribly-named global struct for each overload variant.
* JG: Does it lock us in? Can’t we just forbid them from naming the type?
* BC: That’s an anonymous struct.
* JG: But it’s not so hard to specify. If we think it’s easy enough to add anonymous structs, and it’s not too hard to implement, we should just do that
* MM: So you’d like to have structs named ModF_f32_f32, but authors can’t name those structs?
* JG: yes
* AB: Maybe for MVP, we can leave this as pointers. We spent a lot of time on this, and all the other languages take out parameters
* MM: We could split these functions into two functions, each of which returns one component
* RM: But wouldn’t that be inefficient?
* MM: In HLSL, some of these have to be polyfilled anyway. So it wouldn’t necessarily be a performance loss to make them separate functions
* AB: [mentions a case where there would be performance loss]
* DM: Calling the reserved names out in the spec now is useless. Even if they’re reserved words, new features/extensions would have to add more reserved words, breaking older code.
* JG: We would reserve a prefix, so additions won’t break existing code.
* BC: We already have a reserved prefix in other languages, lile “__xxx”
* DM: (nods)
* AB: There’s a lot of prior art for reserved prefixes.
* JG: even outvars aren’t the end of the world, we can add struct-returning overloads later
* JG: Couldn’t struct returns be post-MVP?
* MM: Having named structs, reserved names or not, is the expedient thing to do at this point.
* RM: Keeping it as-is would be fine for me. It doesn’t require adding anything, it’s ugly but it’s similar to other languages, so it’s not a barrier to adoption
* MM: I surveyed a lot of non-shader languages, and all but one of them use outparams [JB: Did I get that right?]
* MM: These functions are rare, it’s probably not worth having all this discussion.
* AB: Do we want to restrict what storage classes we can use with these outparam pointers?
* DM: In other cases, we’ve just exposed what the back end languages do, but we agreed that we should try to clean things up when the back end languages made mistakes. Now we even have the reserved prefix approach, so why aren’t we making the improvement?
* AB: Because we don’t agree, we have a workable solution for the MVP, and we have a way to fix it post-MVP.
* DM: Does anyone object to returning named structs with a reserved prefix? It seems like everyone liked it
* JG: fatigue?
* RM: I’m happy with that (reserved prefix)
* JG: Okay, let’s do it
* BC: I’ll draft a PR that uses a leading underscore for the struct name
* MM: When people use the word ‘reserved’, that can mean two things:
    * One says you can’t make a new type in that class
    * Another says you can’t even mention things in that class
* BC: We want the second case, so that we can change the names later. I’ll make sure we’re getting the second case when I draft the PR
* consensus \o/


</details>

**Comment by jdashg** [comment]
I believe this is closed by #1973.
